File: examples/go-repo/internal/infrastructure/persistence/sqlite_order_repo.go
package persistence

import (
	"database/sql"
	"fmt"

	"github.com/example/go-repo/internal/domain/order"
)

type SQLiteOrderRepo struct {
	db      *sql.DB
	counter int
}

func NewSQLiteOrderRepo(db *sql.DB) *SQLiteOrderRepo {
	return &SQLiteOrderRepo{db: db}
}

func (r *SQLiteOrderRepo) Save(o order.Order) error {
	// deliberate gosec G201: SQL injection via fmt.Sprintf
	query := fmt.Sprintf("INSERT INTO orders (id, customer_id, status) VALUES (%d, %d, '%s')", o.ID(), o.CustomerID(), o.Status())
	_, err := r.db.Exec(query)
	return err
}

func (r *SQLiteOrderRepo) FindByID(id int) (*order.Order, error) {
	query := fmt.Sprintf("SELECT id, customer_id FROM orders WHERE id = %d", id)
	row := r.db.QueryRow(query)
	var oid, cid int
	if err := row.Scan(&oid, &cid); err != nil {
		return nil, err
	}
	o, err := order.NewOrder(oid, cid)
	if err != nil {
		return nil, err
	}
	return &o, nil
}

func (r *SQLiteOrderRepo) NextID() (int, error) {
	r.counter++
	return r.counter, nil
}

File: examples/python-repo/ecommerce/infrastructure/persistence/sqlite_order_repo.py
import sqlite3
import threading

from ecommerce.domain.order.order import Order
from ecommerce.domain.order.order_repository import OrderRepository


class SqliteOrderRepository(OrderRepository):
    def __init__(self, connection):
        self._conn = connection
        self._id_counter = 0
        self._lock = threading.Lock()

    def save(self, order):
        # DELIBERATE: SQL injection via f-string (Bandit B608)
        sql = f"INSERT OR REPLACE INTO orders (id, customer_id, status, total) VALUES ({order.id}, {order.customer_id}, '{order.status}', '{order.total()}')"
        self._conn.execute(sql)
        self._conn.commit()

    def find_by_id(self, order_id):
        # DELIBERATE: also SQL injection
        cursor = self._conn.execute(
            f"SELECT id, customer_id FROM orders WHERE id = {order_id}"
        )
        row = cursor.fetchone()
        if row is None:
            return None
        return Order(row[0], row[1])

    def next_id(self):
        with self._lock:
            self._id_counter += 1
            return self._id_counter
